yt/frontends/adaptahop/io.py

```diff
@@ -0,0 +1,278 @@
+                    data_file.file_id, pcount)
+        if pcount == 0:
+            return morton
+        ind = 0
+
+        pos = self._get_particle_positions()
+
+        if np.any(pos.min(axis=0) < self.ds.domain_left_edge) or \
+            np.any(pos.max(axis=0) > self.ds.domain_right_edge):
+            raise YTDomainOverflow(pos.min(axis=0),
+                                    pos.max(axis=0),
+                                    self.ds.domain_left_edge,
+                                    self.ds.domain_right_edge)
+        regions.add_data_file(pos, data_file.file_id)
+        morton[ind:ind+pos.shape[0]] = compute_morton(
+            pos[:,0], pos[:,1], pos[:,2],
+            data_file.ds.domain_left_edge,
+            data_file.ds.domain_right_edge)
+
+        return morton
+
+
+    def _count_particles(self, data_file):
+        nhalos = data_file.ds.parameters['nhalos'] + data_file.ds.parameters['nsubs']
+        return {'halos': nhalos}
+
+    def _identify_fields(self, data_file):
+        fields = []
+        for attr, _1, _2 in HALO_ATTRIBUTES:
+            if isinstance(attr, str):
+                fields.append(('halos', attr))
+            else:
+                for a in attr:
+                    fields.append(('halos', a))
+        return fields, {}
+
+    #-----------------------------------------------------
+    # Specific to AdaptaHOP
+    def _get_particle_positions(self):
+        """Read the particles and return them in code_units"""
+        data = getattr(self, '_particle_positions', None)
+        if data is not None:
+            return data
+
+        with FortranFile(self.ds.parameter_filename) as fpu:
+            params = fpu.read_attrs(HEADER_ATTRIBUTES)
+
+            todo = _todo_from_attributes(
+                ('particle_identifier',
+                 'raw_position_x', 'raw_position_y', 'raw_position_z'))
+
+            nhalos = params['nhalos'] + params['nsubs']
+            data = np.zeros((nhalos, 3))
+            offset_map = np.zeros((nhalos, 2), dtype=int)
+            for ihalo in range(nhalos):
+                ipos = fpu.tell()
+                for it in todo:
+                    if isinstance(it, int):
+                        fpu.skip(it)
+                    elif it[0][0] != 'particle_identifier':
+                        # Small optimisation here: we can read as vector
+                        # dt = fpu.read_attrs(it)
+                        # data[ihalo, 0] = dt['particle_position_x']
+                        # data[ihalo, 1] = dt['particle_position_y']
+                        # data[ihalo, 2] = dt['particle_position_z']
+                        data[ihalo, :] = fpu.read_vector(it[0][-1])
+                    else:
+                        halo_id = fpu.read_int()
+                        offset_map[ihalo, 0] = halo_id
+                        offset_map[ihalo, 1] = ipos
+        data = self.ds.arr(data, "code_length") + self.ds.domain_width / 2  # FIXME
```

Comment:
Is there still something that needs to be fixed here?

Reply:
No, it should be good!

Reply:
Ok, just wondering about that #FIXME at the end there. Maybe that should be removed if everything is ok.